File: apps/web/src/coordination.ts
import type { BrowserAccess } from './api';
import type { CollectionExecution } from './collections';
import { isHex } from './graphs';

export interface PackagePin { changeId: string; repositoryId: string; revision: number; digest: string }
export interface RunRepository { repositoryId: string; commit: string; collectionId: string; receiptDigest: string; fullSourceDigest?: string }
export interface RunTask { id: string; perspective: string; profile: string; profileDigest?: string; image?: string; prompt: string; dependsOn: string[]; scopes: {repositoryId: string; writablePaths: string[]}[]; checks: {id: string; repositoryId: string; argv: string[]; timeoutSeconds: number}[]; timeoutSeconds: number }
export interface RunPlan { schemaVersion: number; graphId: string; graphDigest: string; packages: PackagePin[]; repositories: RunRepository[]; tasks: RunTask[]; maxParallel: number }
export interface TaskReceipt { taskId: string; taskKey: string; digest: string; outcome: string; artifactDigest?: string; createdAt: string }
export interface CoordinationRun { id: string; workspaceId: string; repositoryId: string; proposerId: string; plan: RunPlan; digest: string; createdAt: string; authorization?: {actor: string; digest: string; createdAt: string}; cancelRequestedAt?: string; execution?: CollectionExecution; receipts: TaskReceipt[] }
export interface RunPage { runs: {id: string; workspaceId: string; repositoryId: string; digest: string; createdAt: string; authorized: boolean; tasks: number; receipts: number; execution?: CollectionExecution; cancelRequestedAt?: string}[]; nextBefore?: string }
export interface ExecutionCapabilities { repositoryId: string; canExecute: boolean; canPublish: boolean }
export interface ExecutionProfile { id: string; profileDigest: string; image: string; adapter: string; model?: string; command?: string[]; maxBudgetUsd?: string }
export interface ExecutionProfilePage { profiles: ExecutionProfile[]; truncated: boolean }
export function runPath(id: string) { return `/coordination-runs/${encodeURIComponent(id)}`; }
const record = (v: unknown): v is Record<string, unknown> => v !== null && typeof v === 'object' && !Array.isArray(v);
const text = (v: unknown, max: number): v is string => typeof v === 'string' && v.length > 0 && new TextEncoder().encode(v).length <= max && !v.includes('\0') && new TextDecoder().decode(new TextEncoder().encode(v)) === v;
const integer = (v: unknown, low: number, high: number): v is number => Number.isSafeInteger(v) && Number(v) >= low && Number(v) <= high;
const list = (v: unknown, min: number, max: number): v is unknown[] => Array.isArray(v) && v.length >= min && v.length <= max;
const date = (v: unknown) => typeof v === 'string' && Number.isFinite(Date.parse(v));
const execution = (v: unknown) => v === undefined || record(v) && ['namespace','workflowId','runId','state'].every(k => text(v[k],256)) && date(v.observedAt) && typeof v.current === 'boolean';
const keys = (v: Record<string, unknown>, allowed: string[]) => Object.keys(v).every(k => allowed.includes(k));
const image = (v: unknown) => typeof v === 'string' && v.length <= 512 && /^(?:sha256:|[^\s]+@sha256:)[a-f0-9]{64}$/.test(v);

// This validates the preview's shape and bounds. Domain DAG/scope policy and all
// permission decisions remain on the server; a local preview grants no authority.
export function validatePlan(value: unknown): asserts value is RunPlan {
  if (!record(value) || !keys(value,['schemaVersion','graphId','graphDigest','packages','repositories','tasks','maxParallel']) || value.schemaVersion !== 1 || !isHex(value.graphId,32) || !isHex(value.graphDigest,64) || !integer(value.maxParallel,1,4) || !list(value.repositories,1,16) || !list(value.packages,1,16) || !list(value.tasks,1,32) || new TextEncoder().encode(JSON.stringify(value)).length > 1048576) throw new Error('Use a complete version 1 plan with bounded source/package pins and tasks.');
  if (value.repositories.some(r => !record(r) || !keys(r,['repositoryId','commit','collectionId','receiptDigest','fullSourceDigest']) || !text(r.repositoryId,128) || !isHex(r.commit,40) || !isHex(r.collectionId,32) || !isHex(r.receiptDigest,64) || r.fullSourceDigest !== undefined && !isHex(r.fullSourceDigest,64))) throw new Error('Each source must name an exact repository, commit and receipt digest.');
  const repos = new Set(value.repositories.map(r => (r as RunRepository).repositoryId));
  if (repos.size !== value.repositories.length || value.packages.length !== repos.size || value.packages.some(p => !record(p) || !keys(p,['changeId','repositoryId','revision','digest']) || !text(p.changeId,128) || !repos.has(String(p.repositoryId)) || !integer(p.revision,1,Number.MAX_SAFE_INTEGER) || !isHex(p.digest,64)) || new Set(value.packages.map(p=>(p as PackagePin).repositoryId)).size!==repos.size) throw new Error('Include one exact package revision per source repository.');
  if (value.tasks.some(t => !record(t) || !keys(t,['id','perspective','profile','profileDigest','image','prompt','dependsOn','scopes','checks','timeoutSeconds']) || !text(t.id,64) || !text(t.profile,64) || !['architect','developer','qc','product','operations'].includes(String(t.perspective)) || !text(t.prompt,32768) || !integer(t.timeoutSeconds,1,1800) || t.profileDigest!==undefined&&!isHex(t.profileDigest,64) || t.image!==undefined&&!image(t.image)
    || !list(t.dependsOn ?? [],0,32) || (t.dependsOn as unknown[]|undefined)?.some(d=>!text(d,64)) || !list(t.scopes,1,16)
    || t.scopes.some(s=>!record(s)||!keys(s,['repositoryId','writablePaths'])||!repos.has(String(s.repositoryId))||!list(s.writablePaths ?? [],0,128)||(s.writablePaths as unknown[]|undefined)?.some(p=>!text(p,1024)))
    || !list(t.checks ?? [],0,16) || (t.checks as unknown[]|undefined)?.some(c=>!record(c)||!keys(c,['id','repositoryId','argv','timeoutSeconds'])||!text(c.id,64)||!repos.has(String(c.repositoryId))||!list(c.argv,1,64)||c.argv.some(a=>typeof a!=='string'||new TextEncoder().encode(a).length>4096||a.includes('\0'))||!integer(c.timeoutSeconds,1,1800)))) throw new Error('A task has incomplete scope, profile, dependencies or verification commands.');
  const tasks=value.tasks as unknown as RunTask[];
  if(new Set(tasks.map(t=>t.id)).size!==tasks.length)throw new Error('Task identifiers must be unique.');
}
export function normalizePlan(plan: RunPlan): RunPlan { return {...plan,tasks:plan.tasks.map(t=>({...t,dependsOn:t.dependsOn??[],checks:t.checks??[],scopes:t.scopes.map(s=>({...s,writablePaths:s.writablePaths??[]}))}))}; }
export function validateRun(run: CoordinationRun, access: BrowserAccess, id?: string) {
  if (!run || !isHex(run.id,32) || id!==undefined&&run.id!==id || run.workspaceId!==access.workspaceID || run.repositoryId!==access.repositoryID || !isHex(run.digest,64) || !text(run.proposerId,128) || !date(run.createdAt) || !execution(run.execution)) throw new Error('This run does not match the selected workspace and repository.');
  validatePlan(run.plan);
  if(!run.plan.repositories.some(r=>r.repositoryId===access.repositoryID) || !list(run.receipts,0,32) || run.receipts.some(r=>!r||!isHex(r.taskId,32)||!text(r.taskKey,64)||!run.plan.tasks.some(t=>t.id===r.taskKey)||!isHex(r.digest,64)||!text(r.outcome,64)||r.artifactDigest!==undefined&&!isHex(r.artifactDigest,64)||!date(r.createdAt)) || new Set(run.receipts.map(r=>r.taskKey)).size!==run.receipts.length || run.authorization!==undefined&&(!run.authorization||run.authorization.digest!==run.digest||!text(run.authorization.actor,128)||!date(run.authorization.createdAt)) || run.cancelRequestedAt!==undefined&&!date(run.cancelRequestedAt)) throw new Error('The run receipt or authorization response is incomplete.');
}
export function validateRunPage(page: RunPage, access: BrowserAccess) {
  if(!page||!list(page.runs,0,20)||page.runs.some(r=>!r||!isHex(r.id,32)||!isHex(r.digest,64)||r.workspaceId!==access.workspaceID||r.repositoryId!==access.repositoryID||!date(r.createdAt)||!execution(r.execution)||typeof r.authorized!=='boolean'||!integer(r.tasks,1,32)||!integer(r.receipts,0,32))||new Set(page.runs.map(r=>r.id)).size!==page.runs.length||page.nextBefore!==undefined&&!text(page.nextBefore,1024))throw new Error('The shared run list is incomplete. Refresh before inspecting a run.');
}
export function validateProfiles(page: ExecutionProfilePage) {
  if(!page||!list(page.profiles,0,100)||typeof page.truncated!=='boolean'||page.profiles.some(p=>!p||!text(p.id,64)||!isHex(p.profileDigest,64)||!image(p.image)||!text(p.adapter,64))||new Set(page.profiles.map(p=>p.id)).size!==page.profiles.length)throw new Error('The execution profile catalog is incomplete.');
}

File: internal/domain/coordination.go
package domain

import (
	"encoding/json"
	"fmt"
	"sort"
	"strings"
	"time"
	"unicode/utf8"
)

const MaxCoordinationTasks = 32

type ExecutionCapabilities struct {
	RepositoryID string `json:"repositoryId"`
	CanExecute   bool   `json:"canExecute"`
	CanPublish   bool   `json:"canPublish"`
}

// Execution grants are provisioned independently from design-review grants.
// A person may review architecture without being permitted to spend execution
// resources or publish code. Agents cannot turn either grant into human authority.
type ExecutionGrantConfig struct {
	RepositoryID string `json:"repositoryId"`
	PrincipalID  string `json:"principalId"`
	CanExecute   bool   `json:"canExecute"`
	CanPublish   bool   `json:"canPublish"`
}

type PackagePin struct {
	ChangeID     string `json:"changeId"`
	RepositoryID string `json:"repositoryId"`
	Revision     int64  `json:"revision"`
	Digest       string `json:"digest"`
}

type CoordinationRepository struct {
	RepositoryID     string `json:"repositoryId"`
	Commit           string `json:"commit"`
	CollectionID     string `json:"collectionId"`
	ReceiptDigest    string `json:"receiptDigest"`
	FullSourceDigest string `json:"fullSourceDigest,omitempty"`
}

type TaskScope struct {
	RepositoryID  string   `json:"repositoryId"`
	WritablePaths []string `json:"writablePaths"`
}

type VerificationCommand struct {
	ID             string                    `json:"id"`
	RepositoryID   string                    `json:"repositoryId"`
	Argv           []string                  `json:"argv"`
	TimeoutSeconds int                       `json:"timeoutSeconds"`
	Requirements   []VerificationRequirement `json:"requirements,omitempty"`
}

type CoordinationTask struct {
	ID             string                `json:"id"`
	Perspective    string                `json:"perspective"`
	Profile        string                `json:"profile"`
	ProfileDigest  string                `json:"profileDigest,omitempty"`
	Image          string                `json:"image,omitempty"`
	Prompt         string                `json:"prompt"`
	DependsOn      []string              `json:"dependsOn"`
	Scopes         []TaskScope           `json:"scopes"`
	Checks         []VerificationCommand `json:"checks"`
	TimeoutSeconds int                   `json:"timeoutSeconds"`
}

// The immutable plan binds every producer to inspected packages and baselines.
// Task IDs are stable plan keys; workflow payloads use separate server-generated
// task IDs so prompts, source and review content stay outside workflow history.
type CoordinationPlan struct {
	SchemaVersion int                      `json:"schemaVersion"`
	GraphID       string                   `json:"graphId"`
	GraphDigest   string                   `json:"graphDigest"`
	Packages      []PackagePin             `json:"packages"`
	Repositories  []CoordinationRepository `json:"repositories"`
	Tasks         []CoordinationTask       `json:"tasks"`
	MaxParallel   int                      `json:"maxParallel"`
}

type RunAuthorization struct {
	Actor     string    `json:"actor"`
	Digest    string    `json:"digest"`
	CreatedAt time.Time `json:"createdAt"`
}

type TaskReceipt struct {
	TaskID         string    `json:"taskId"`
	TaskKey        string    `json:"taskKey"`
	Digest         string    `json:"digest"`
	Outcome        string    `json:"outcome"`
	ArtifactDigest string    `json:"artifactDigest,omitempty"`
	CreatedAt      time.Time `json:"createdAt"`
}

type CoordinationRun struct {
	ID                string               `json:"id"`
	WorkspaceID       string               `json:"workspaceId"`
	RepositoryID      string               `json:"repositoryId"`
	ProposerID        string               `json:"proposerId"`
	Plan              CoordinationPlan     `json:"plan"`
	Digest            string               `json:"digest"`
	CreatedAt         time.Time            `json:"createdAt"`
	Authorization     *RunAuthorization    `json:"authorization,omitempty"`
	CancelRequestedAt *time.Time           `json:"cancelRequestedAt,omitempty"`
	Execution         *CollectionExecution `json:"execution,omitempty"`
	Receipts          []TaskReceipt        `json:"receipts"`
}

type CoordinationPage struct {
	Runs       []CoordinationSummary `json:"runs"`
	NextBefore string                `json:"nextBefore,omitempty"`
}

// Discovery stays bounded even when a plan contains many prompts and paths.
// Retrieve an individual run to inspect its exact plan before authorization.
type CoordinationSummary struct {
	ID                string               `json:"id"`
	WorkspaceID       string               `json:"workspaceId"`
	RepositoryID      string               `json:"repositoryId"`
	ProposerID        string               `json:"proposerId"`
	Digest            string               `json:"digest"`
	CreatedAt         time.Time            `json:"createdAt"`
	Authorized        bool                 `json:"authorized"`
	CancelRequestedAt *time.Time           `json:"cancelRequestedAt,omitempty"`
	Execution         *CollectionExecution `json:"execution,omitempty"`
	Tasks             int                  `json:"tasks"`
	Receipts          int                  `json:"receipts"`
}

func planKey(s string) bool {
	if len(s) < 1 || len(s) > 64 {
		return false
	}
	for _, c := range s {
		if (c < 'a' || c > 'z') && (c < 'A' || c > 'Z') && (c < '0' || c > '9') && c != '-' && c != '_' && c != '.' {
			return false
		}
	}
	return true
}

func scopedPath(path string) bool {
	if path == "" || len(path) > 1024 || !utf8.ValidString(path) || strings.TrimSpace(path) == "" || strings.HasPrefix(path, "/") || strings.HasPrefix(path, "-") || strings.ContainsAny(path, "\\\x00\r\n\t") {
		return false
	}
	for _, c := range path {
		if c < 32 || c == 127 {
			return false
		}
	}
	for _, part := range strings.Split(path, "/") {
		if part == "" || part == "." || part == ".." || strings.EqualFold(part, ".git") {
			return false
		}
	}
	return true
}

// ExecutionImagePinned accepts only immutable image identities reviewed in a plan.
func ExecutionImagePinned(image string) bool {
	if strings.HasPrefix(image, "sha256:") {
		return IsLowerHex(strings.TrimPrefix(image, "sha256:"), 64)
	}
	parts := strings.Split(image, "@sha256:")
	return len(parts) == 2 && parts[0] != "" && !strings.ContainsAny(parts[0], " \t\r\n") && IsLowerHex(parts[1], 64)
}

func PathsOverlap(a, b string) bool {
	return a == b || strings.HasPrefix(a, b+"/") || strings.HasPrefix(b, a+"/")
}

// Empty task collections have one retained representation, so omitted optional
// lists and explicit empty lists do not fork an idempotent proposal's identity.
func NormalizeCoordinationPlan(p CoordinationPlan) (CoordinationPlan, error) {
	if err := ValidateCoordinationPlan(p); err != nil {
		return CoordinationPlan{}, err
	}
	p.Tasks = append([]CoordinationTask{}, p.Tasks...)
	for i := range p.Tasks {
		t := &p.Tasks[i]
		t.DependsOn = append([]string{}, t.DependsOn...)
		t.Checks = append([]VerificationCommand{}, t.Checks...)
		for j := range t.Checks {
			t.Checks[j].Requirements = append([]VerificationRequirement(nil), t.Checks[j].Requirements...)
		}
		t.Scopes = append([]TaskScope{}, t.Scopes...)
		for j := range t.Scopes {
			t.Scopes[j].WritablePaths = append([]string{}, t.Scopes[j].WritablePaths...)
		}
	}
	return p, nil
}

// ValidateCoordinationPlan rejects ambiguous ownership before admission. Tasks
// writing overlapping paths must be ordered by an explicit dependency, rather
// than relying on model cooperation or which parallel worker happens to finish.
func ValidateCoordinationPlan(p CoordinationPlan) error {
	invalid := func(why string) error { return fmt.Errorf("%w: %s", ErrInvalidInput, why) }
	encoded, err := json.Marshal(p)
	if err != nil || len(encoded) > 1<<20 {
		return invalid("plan exceeds encoded size bound")
	}
	if p.SchemaVersion != 1 || !IsLowerHex(p.GraphID, 32) || !IsLowerHex(p.GraphDigest, 64) || p.MaxParallel < 1 || p.MaxParallel > 4 || len(p.Repositories) < 1 || len(p.Repositories) > 16 || len(p.Packages) != len(p.Repositories) || len(p.Tasks) < 1 || len(p.Tasks) > MaxCoordinationTasks {
		return invalid("invalid plan identity or bounds")
	}
	repos := map[string]bool{}
	for _, r := range p.Repositories {
		if r.FullSourceDigest != "" && !IsLowerHex(r.FullSourceDigest, 64) {
			return ErrInvalidInput
		}
		if ValidateAccessID(r.RepositoryID) != nil || repos[r.RepositoryID] || !IsLowerHex(r.Commit, 40) || !IsLowerHex(r.CollectionID, 32) || !IsLowerHex(r.ReceiptDigest, 64) {
			return invalid("each repository needs one exact source receipt")
		}
		repos[r.RepositoryID] = true
	}
	packages := map[string]bool{}
	pinned := map[string]bool{}
	for _, pin := range p.Packages {
		if ValidateAccessID(pin.ChangeID) != nil || !repos[pin.RepositoryID] || pinned[pin.RepositoryID] || packages[pin.ChangeID] || pin.Revision < 1 || !IsLowerHex(pin.Digest, 64) {
			return invalid("each repository needs one exact package revision")
		}
		pinned[pin.RepositoryID] = true
		packages[pin.ChangeID] = true
	}
	tasks := map[string]CoordinationTask{}
	for _, task := range p.Tasks {
		if task.ProfileDigest != "" && !IsLowerHex(task.ProfileDigest, 64) || task.Image != "" && !ExecutionImagePinned(task.Image) {
			return ErrInvalidInput
		}
		if !planKey(task.ID) || !planKey(task.Profile) || task.Prompt == "" || len(task.Prompt) > 32<<10 || !utf8.ValidString(task.Prompt) || len(task.Scopes) < 1 || len(task.Scopes) > len(repos) || len(task.Checks) > 16 || len(task.DependsOn) > MaxCoordinationTasks || task.TimeoutSeconds < 1 || task.TimeoutSeconds > 1800 {
			return invalid("invalid task bounds or identity")
		}
		if _, ok := tasks[task.ID]; ok {
			return invalid("duplicate task ID")
		}
		switch task.Perspective {
		case "architect", "developer", "qc", "product", "operations":
		default:
			return invalid("unknown task perspective")
		}
		scopes := map[string]bool{}
		for _, scope := range task.Scopes {
			if !repos[scope.RepositoryID] || scopes[scope.RepositoryID] || len(scope.WritablePaths) > 128 {
				return invalid("invalid task repository scope")
			}
			scopes[scope.RepositoryID] = true
			paths := map[string]bool{}
			for _, path := range scope.WritablePaths {
				if !scopedPath(path) || paths[path] {
					return invalid("invalid or duplicate writable path")
				}
				paths[path] = true
			}
		}
		checks := map[string]bool{}
		for _, check := range task.Checks {
			if ValidateVerificationRequirements(check.Requirements) != nil {
				return invalid("invalid verification requirement references")
			}
			for _, requirement := range check.Requirements {
				matched := false
				for _, pin := range p.Packages {
					if pin.ChangeID == requirement.ChangeID && pin.Revision == requirement.Revision && pin.Digest == requirement.Digest && scopes[pin.RepositoryID] {
						matched = true
						break
					}
				}
				if !matched {
					return invalid("verification requirement must match an exact task-scoped package pin")
				}
			}
			if !planKey(check.ID) || checks[check.ID] || !scopes[check.RepositoryID] || len(check.Argv) < 1 || len(check.Argv) > 64 || check.TimeoutSeconds < 1 || check.TimeoutSeconds > task.TimeoutSeconds {
				return invalid("invalid verification command")
			}
			checks[check.ID] = true
			for _, arg := range check.Argv {
				if len(arg) > 4096 || !utf8.ValidString(arg) || strings.ContainsRune(arg, 0) {
					return invalid("invalid verification argument")
				}
			}
			if strings.TrimSpace(check.Argv[0]) == "" {
				return invalid("verification executable required")
			}
		}
		tasks[task.ID] = task
	}
	marks := map[string]int{}
	ancestors := map[string]map[string]bool{}
	var visit func(string) error
	visit = func(id string) error {
		if marks[id] == 1 {
			return invalid("task dependencies form a cycle")
		}
		if marks[id] == 2 {
			return nil
		}
		marks[id] = 1
		task, ok := tasks[id]
		if !ok {
			return invalid("task dependency does not exist")
		}
		ancestors[id] = map[string]bool{}
		seen := map[string]bool{}
		for _, dep := range task.DependsOn {
			if dep == id || seen[dep] {
				return invalid("duplicate or self dependency")
			}
			seen[dep] = true
			if err := visit(dep); err != nil {
				return err
			}
			ancestors[id][dep] = true
			for ancestor := range ancestors[dep] {
				ancestors[id][ancestor] = true
			}
		}
		marks[id] = 2
		return nil
	}
	for _, task := range p.Tasks {
		if err := visit(task.ID); err != nil {
			return err
		}
	}
	for i, a := range p.Tasks {
		for _, b := range p.Tasks[i+1:] {
			if ancestors[a.ID][b.ID] || ancestors[b.ID][a.ID] {
				continue
			}
			for _, as := range a.Scopes {
				for _, bs := range b.Scopes {
					if as.RepositoryID != bs.RepositoryID {
						continue
					}
					for _, ap := range as.WritablePaths {
						for _, bp := range bs.WritablePaths {
							if PathsOverlap(ap, bp) {
								return invalid("parallel tasks have overlapping write scopes")
							}
						}
					}
				}
			}
		}
	}
	data, err := JSONDigest(p)
	if err != nil || data == "" {
		return invalid("invalid plan encoding")
	}
	return nil
}

func CoordinationRepositoryIDs(plan CoordinationPlan) []string {
	ids := make([]string, 0, len(plan.Repositories))
	for _, r := range plan.Repositories {
		ids = append(ids, r.RepositoryID)
	}
	sort.Strings(ids)
	return ids
}
